### DISCUSSION TITLE
Review Logic Seems Flawed — Nevermind...

### DISCUSSION THREAD

**Discussion by johnmarksilly** [blakeblackshear/frigate#14971] [Ask A Question] [answered]
So here is my biggest gripe with the review mechanism. Before I could ignore all events that weren't on my property through a label, but now all events get grouped together with property labeled events if they occur with close timestamps. 

I have a camera facing an intersection right in front of my house. I want to record the intersection, but I don't care about most events. 

I mark the config, alerts -> required_zones: property. I don't care if detections contain non-property events. 

Unfortunately though, when this intersection gets busy and a car does come into my driveway, every subsequent car that passes by re-triggers an update to this review item. Attached is an example. 

Only one car is actually on my property, but because all the other cars happened to drive by within short time periods of each other, they re-trigger the alert. 

![image](https://github.com/user-attachments/assets/687446dc-230c-4158-b8ea-6c6b26af7f8e)

Everything further down is a vehicle passing by. 

Any suggestions?

Basically my only thought is to now remove cars as a label for alerts, but even people will be an issue since that intersection is a bus pickup.

**Comment by NickM-27** [comment]
Just remove car from the detections -> label list

**Reply by johnmarksilly** [reply]
The problem with removing car is that I no longer will see the snapshot in the detections section. 

My scenario is as follows, all alerts should be labels I care about that enter my property zone. 

These are the only things I want alerts for. 

However, just recently I had a situation where I needed to see if  I got a plate for a particular car. So I needed cars as a label in the detections category in order to see the ones not on my property. 

I have 24/7 recordings of everything, but it's not helpful to scrub through hours of footage. Now that individual events aren't surfaced up, I don't see those without having the detection set up in review. 

I'm not sure I understand why setting required_zones doesn't force review items to actually require that zone. It basically just means, required_start_zones. 

There should be a configurable cooldown that I could set to 5 seconds or something low or be able to force all events to also be of a zone before aggregating them together.

When the feature was rolled out, I thought it would aggregate across cameras too which would be ideal.

**Reply by NickM-27** [reply]
> I'm not sure I understand why setting required_zones doesn't force review items to actually require that zone. It basically just means, required_start_zones.

If someone came and stole packages of your porch then walked to the next house and did the same that would all be part of one activity. It doesn't make sense to only include the time when the object is in the zone because you lose context for that objects activity.

> There should be a configurable cooldown that I could set to 5 seconds or something low or be able to force all events to also be of a zone before aggregating them together.

There is a cool down, I suppose it could be made configurable but if it's too low you'll find times where objects are in and out of frame (like unloading a car) to be put into separate review items

**Reply by johnmarksilly** [reply]
That is obviously the use-case and is the preferred way of handling events, but that's in a perfect world where only those events are the ones that are captured.

In the real world, that happens much less frequently than the common scenario; packages or people arrive, do their thing, and leave.

In any case, the documentation should be revised to be very specific about that setting and potentially show a sequence diagram. The `required_zones` field isn't descriptive enough because in the past, for events, that meant no event would be created without it occurring within the zone.

> There is a cool down, I suppose it could be made configurable but if it's too low you'll find times where objects are in and out of frame (like unloading a car) to be put into separate review items

I think I would prefer the old event system to the review system because of this exact flaw. I like the idea of bundling the events together and I think it provides the best UX (even bundling events across cameras together), but in its current state, it is less than ideal. The only way it works is if a model is trained to correlate labels across events.

**Reply by NickM-27** [reply]
> In the real world, that happens much less frequently than the common scenario; packages or people arrive, do their thing, and leave.

We have not received any feedback like this and even in my case living on a busy street I don't have this issue so I would say it is not that common. 

> I think I would prefer the old event system to the review system because of this exact flaw. I like the idea of bundling the events together and I think it provides the best UX (even bundling events across cameras together), but in its current state, it is less than ideal. 

The explore view pretty much covers that in 0.15, I don't foresee any major changes happening to the review system. I don't think it would be that difficult to have some logic that understands when an alert causing object is gone and end the current review item and immediately start a new one, but that is just more complicated and not something that anymore reported during the beta or release thus far. 

> The only way it works is if a model is trained to correlate labels across events.

Not sure what you mean, an "event" (tracked object) is the same throughout the time that object is tracked.

**Comment by johnmarksilly** [comment]
First off, holy shit… I’ve never even clicked on the explore view. This is exactly what I was thinking. I’m so sorry.

Second this is all anecdotal right? The review system is still pretty young so either way it would be difficult to understand usage.

So, let me demonstrate this issue right here:

I have an event right now in my Alerts from a car arriving on my driveway (wife came home). This occurred at 7:10 PM last night. I assume because it is stationary and still on my property the event continues. All night, cars have been added to this event. I think there is a total of 6 cars listed in the review item. My wife left our house in the same car at 7:15 AM and it is in the exact same review item! That’s 12 hours! This should not have kept going that long. 

￼

**Reply by NickM-27** [reply]
> Second this is all anecdotal right? The review system is still pretty young so either way it would be difficult to understand usage.

we spent 8 months developing the new UI throughout which we had ~20 users using the dev images and giving feedback and also 4 months of that was a beta period where we had ~ 5000 users using the beta many of which gave feedback (multiple discussions with 1000s of comments). Given the issues that were caught during this period and the immense feedback that we have received, I disagree that its usage is difficult to understand. 

> I have an event right now in my Alerts from a car arriving on my driveway (wife came home). This occurred at 7:10 PM last night. I assume because it is stationary and still on my property the event continues. All night, cars have been added to this event. I think there is a total of 6 cars listed in the review item. My wife left our house in the same car at 7:15 AM and it is in the exact same review item! That’s 12 hours! This should not have kept going that long.

stationary objects will not keep a review item going, as they are only periods of activity. Sounds like something is wrong with the config or some other issue is occurring.

**Reply by johnmarksilly** [reply]
I’m feeling a bit confused about the review system. I’m not criticizing it, I just want to understand how it works. I also think the alerts should be shorter. Maybe I should have started with a more friendly message. Looking back it looks a little combative.

This is my config for that camera:
```
  southwest:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/southwest
          roles:
            - record
            - detect
    motion:
      threshold: 35
      mask:
        - 0.834,0.936,0.835,0.976,0.992,0.98,0.991,0.936
        - 0,0.228,0.1,0.224,0.193,0.169,0.383,0.093,0.465,0.082,0.461,0.138,0.657,0.154,0.773,0.177,0.927,0.223,1,0.248,1,0,0,0
      contour_area: 10
      improve_contrast: true
    zones:
      property:
        coordinates: 
          0,0.555,0.076,0.458,0.154,0.404,0.24,0.386,0.469,0.406,0.719,0.421,0.939,0.439,1,0.44,1,1,0,1
        inertia: 3
        loitering_time: 0
      property_edge:
        coordinates: 
          0,0.456,0.068,0.392,0.119,0.366,0.185,0.33,0.256,0.325,0.462,0.318,0.763,0.339,0.981,0.384,1,0.404,1,0.44,0.258,0.387,0.158,0.404,0.071,0.463,0,0.554
        inertia: 10
        loitering_time: 0
    review:
      alerts:
        required_zones: property
```

I don't think it is anything out of the ordinary. Does the review item continue to aggregate based solely on motion? I think it has to be a bug because on that particular camera, there were probably 30 more car detections that I can see on that camera over the same time period that went into the event.

**Reply by johnmarksilly** [reply]
![image](https://github.com/user-attachments/assets/82793773-a983-406d-ac04-3c71ee8c7984)

**Reply by NickM-27** [reply]
> Does the review item continue to aggregate based solely on motion?

A review item will only continue as long as there is a non stationary (active) object detected on that camera, OR if you have a zone with loitering time it will continue as long as an object is inside of the loitering zone. 

In this case my only guess is there was some car that was not detected well so the bounding box was bouncing and not consistent enough for the car to be considered stationary.

**Reply by johnmarksilly** [reply]
Hmm.. But if there was something keeping it active, why wouldn't the other vehicles that were detected driving on the road show up in the same review item?

I'll make sure to take a look at the debug view later tonight and get back to you.

**Reply by NickM-27** [reply]
> Hmm.. But if there was something keeping it active, why wouldn't the other vehicles that were detected driving on the road show up in the same review item?

Based on what you said they do show up in the review item menu. Might need more screenshots to understand what you're talking about

**Reply by johnmarksilly** [reply]
Here:
![image](https://github.com/user-attachments/assets/580219d8-8dee-448d-af87-29fac271b5eb)

These are some events before my wife left from the same camera (that top one is one of the same events that showed up in the alert).

There was probably an additional 30 or so car events that showed up on the camera between the time the alert started and the time it ended.

**Reply by NickM-27** [reply]
Need to see the metadata for this long running review item then, because realistically it's not possible for multiple review items to happen during the same time period on the same camera

**Reply by johnmarksilly** [reply]
What would you need me to pull? Should I grab the response of the review id? Is there a more detailed API response?

**Reply by NickM-27** [reply]
Yeah the api response would be fine

**Reply by johnmarksilly** [reply]
Not super interesting
```
{
  "id": "1731553606.374358-zdhzd5",
  "camera": "southwest",
  "start_time": 1731553606.374358,
  "end_time": 1731553759.024454,
  "has_been_reviewed": true,
  "severity": "alert",
  "thumb_path": "/media/frigate/clips/review/thumb-southwest-1731553606.374358-zdhzd5.webp",
  "data": {
    "detections": [
      "1731553632.103296-4ncdkm",
      "1731553629.723537-yohjqr",
      "1731553609.296109-ks6p0z",
      "1731553605.493978-0j4p3k",
      "1731553681.500203-1tlp9u"
    ],
    "objects": [
      "car"
    ],
    "sub_labels": [],
    "zones": [
      "property",
      "property_edge"
    ],
    "audio": []
  }
```

Here is the car on my driveway (ends before the review item ends):
```
{
  "id": "1731553632.103296-4ncdkm",
  "label": "car",
  "sub_label": null,
  "camera": "southwest",
  "start_time": 1731553632.103296,
  "end_time": 1731553640.189966,
  "top_score": null,
  "score": null,
  "false_positive": null,
  "zones": [],
  "thumbnail": "xxxx",
  "has_clip": true,
  "has_snapshot": true,
  "region": null,
  "box": null,
  "area": null,
  "retain_indefinitely": false,
  "ratio": 1.0,
  "plus_id": null,
  "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
  "detector_type": "edgetpu",
  "model_type": "ssd",
  "data": {
    "box": [
      0.05703125,
      0.24444444444444444,
      0.06953125,
      0.0787037037037037
    ],
    "region": [
      0.0,
      0.11481481481481481,
      0.225,
      0.4
    ],
    "score": 0.8828125,
    "top_score": 0.8984375,
    "attributes": [],
    "type": "object"
  }
}
```

**Reply by NickM-27** [reply]
> 1731553759.024454 - 1731553606.374358 = 152.65

this review item is 152 seconds long, that is not overnight, so I think something is being misunderstood here

**Reply by johnmarksilly** [reply]
Sorry that was the wrong one. 
```
{
  "id": "1731546637.707816-a7j7ef",
  "camera": "southwest",
  "start_time": 1731546637.707816,
  "end_time": 1731547018.579317,
  "has_been_reviewed": true,
  "severity": "alert",
  "thumb_path": "/media/frigate/clips/review/thumb-southwest-1731546637.707816-a7j7ef.webp",
  "data": {
    "detections": [
      "1731546635.562671-osimic",
      "1731546747.374943-g8paey",
      "1731546530.228501-6nc1q7",
      "1731546741.01269-8oi1kt",
      "1731547015.779655-a31idn",
      "1731546766.564828-cbysf8"
    ],
    "objects": [
      "car"
    ],
    "sub_labels": [],
    "zones": [
      "property"
    ],
    "audio": []
  }
}
```

whole event list:
```
[{
    "id": "1731546530.228501-6nc1q7",
    "label": "car",
    "sub_label": null,
    "camera": "southwest",
    "start_time": 1731546530.228501,
    "end_time": 1731546752.81693,
    "top_score": null,
    "score": null,
    "false_positive": null,
    "zones": ["property"],
    "thumbnail": "xxx",
    "has_clip": true,
    "has_snapshot": true,
    "region": null,
    "box": null,
    "area": null,
    "retain_indefinitely": false,
    "ratio": 1.0,
    "plus_id": null,
    "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
    "detector_type": "edgetpu",
    "model_type": "ssd",
    "data": {
        "box": [0.9033854166666667, 0.3638888888888889, 0.0953125, 0.2574074074074074],
        "region": [0.60625, 0.025, 0.39375, 0.7],
        "score": 0.83203125,
        "top_score": 0.85546875,
        "attributes": [],
        "type": "object"
    }
}, {
    "id": "1731546635.562671-osimic",
    "label": "car",
    "sub_label": null,
    "camera": "southwest",
    "start_time": 1731546635.562671,
    "end_time": 1731546991.362765,
    "top_score": null,
    "score": null,
    "false_positive": null,
    "zones": [],
    "thumbnail": "xxx",
    "has_clip": true,
    "has_snapshot": true,
    "region": null,
    "box": null,
    "area": null,
    "retain_indefinitely": false,
    "ratio": 1.0,
    "plus_id": null,
    "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
    "detector_type": "edgetpu",
    "model_type": "ssd",
    "data": {
        "box": [0.034895833333333334, 0.2310185185185185, 0.037760416666666664, 0.043055555555555555],
        "region": [0.011458333333333333, 0.17962962962962964, 0.08333333333333333, 0.14814814814814814],
        "score": 0.90625,
        "top_score": 0.90234375,
        "attributes": [{
            "box": [0.0375, 0.25925925925925924, 0.003125, 0.005555555555555556],
            "label": "license_plate",
            "score": 0.60546875
        }],
        "type": "object"
    }
}, {
    "id": "1731546741.01269-8oi1kt",
    "label": "car",
    "sub_label": null,
    "camera": "southwest",
    "start_time": 1731546741.01269,
    "end_time": null,
    "top_score": null,
    "score": null,
    "false_positive": null,
    "zones": [],
    "thumbnail": "xxx",
    "has_clip": true,
    "has_snapshot": true,
    "region": null,
    "box": null,
    "area": null,
    "retain_indefinitely": false,
    "ratio": 1.0,
    "plus_id": null,
    "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
    "detector_type": "edgetpu",
    "model_type": "ssd",
    "data": {
        "box": [0.9169270833333333, 0.20601851851851852, 0.04609375, 0.06157407407407407],
        "region": [0.8979166666666667, 0.16296296296296298, 0.08333333333333333, 0.14814814814814814],
        "score": 0.98046875,
        "top_score": 0.984375,
        "attributes": [{
            "box": [0.95078125, 0.23935185185185184, 0.00703125, 0.008333333333333333],
            "label": "license_plate",
            "score": 0.60546875
        }],
        "type": "object"
    }
}, {
    "id": "1731546747.374943-g8paey",
    "label": "car",
    "sub_label": null,
    "camera": "southwest",
    "start_time": 1731546747.374943,
    "end_time": 1731547508.017024,
    "top_score": null,
    "score": null,
    "false_positive": null,
    "zones": ["property"],
    "thumbnail": "xxx",
    "has_clip": true,
    "has_snapshot": true,
    "region": null,
    "box": null,
    "area": null,
    "retain_indefinitely": false,
    "ratio": 1.0,
    "plus_id": null,
    "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
    "detector_type": "edgetpu",
    "model_type": "ssd",
    "data": {
        "box": [0.7111979166666667, 0.2101851851851852, 0.1203125, 0.09583333333333334],
        "region": [0.6830729166666667, 0.12453703703703704, 0.23645833333333333, 0.4203703703703704],
        "score": 0.8828125,
        "top_score": 0.890625,
        "attributes": [],
        "type": "object"
    }
}, {
    "id": "1731546766.564828-cbysf8",
    "label": "car",
    "sub_label": null,
    "camera": "southwest",
    "start_time": 1731546766.564828,
    "end_time": 1731546771.60668,
    "top_score": null,
    "score": null,
    "false_positive": null,
    "zones": [],
    "thumbnail": "xxx",
    "has_clip": true,
    "has_snapshot": true,
    "region": null,
    "box": null,
    "area": null,
    "retain_indefinitely": false,
    "ratio": 1.0,
    "plus_id": null,
    "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
    "detector_type": "edgetpu",
    "model_type": "ssd",
    "data": {
        "box": [0.025, 0.25601851851851853, 0.10494791666666667, 0.11666666666666667],
        "region": [0.0, 0.1300925925925926, 0.246875, 0.4388888888888889],
        "score": 0.92578125,
        "top_score": 0.857421875,
        "attributes": [],
        "type": "object"
    }
}, {
    "id": "1731547015.779655-a31idn",
    "label": "car",
    "sub_label": null,
    "camera": "southwest",
    "start_time": 1731547015.779655,
    "end_time": 1731547019.08458,
    "top_score": null,
    "score": null,
    "false_positive": null,
    "zones": [],
    "thumbnail": "xxx",
    "has_clip": true,
    "has_snapshot": true,
    "region": null,
    "box": null,
    "area": null,
    "retain_indefinitely": false,
    "ratio": 1.0,
    "plus_id": null,
    "model_hash": "d5cb985ddaa212dcdbf81cfc41c8d9c6",
    "detector_type": "edgetpu",
    "model_type": "ssd",
    "data": {
        "box": [0.030729166666666665, 0.2462962962962963, 0.09635416666666667, 0.12175925925925926],
        "region": [0.029166666666666667, 0.18611111111111112, 0.121875, 0.21666666666666667],
        "score": 0.890625,
        "top_score": 0.861328125,
        "attributes": [],
        "type": "object"
    }
}]
```

**Reply by NickM-27** [reply]
That one is still only 380 seconds long

**Reply by johnmarksilly** [reply]
yeah I know that's why it's weird, but if you look at the events, there is one that doesn't have an end time.

**Reply by NickM-27** [reply]
> yeah I know that's why it's weird, but if you look at the events, there is one that doesn't have an end time.

where are you looking at "events" exactly?

**Reply by NickM-27** [reply]
okay, if you are referring to the events as in the events enpoint then that is expected because events are just tracked objects and your car is still there hence the object does not have an end time. But events are not used for the review items UI so it is not clear why these two things are being conflated.

**Reply by johnmarksilly** [reply]
When the UI gets the thumbnails, it makes a call to grab all of the event details:
`http://hawkeye.surveillance:5000/api/event_ids?ids=1731546635.562671-osimic,1731546747.374943-g8paey,1731546530.228501-6nc1q7,1731546741.01269-8oi1kt,1731547015.779655-a31idn,1731546766.564828-cbysf8`

**Reply by NickM-27** [reply, marked as answer]
right, but that doesn't affect the review items. 

The original statement was

> My wife left our house in the same car at 7:15 AM and it is in the exact same review item! That’s 12 hours! This should not have kept going that long.

but so far there is nothing showing that the review item lasted that long or anywhere close, only the individual event for that car (which is behaving as expected, is exactly how events behaved in 0.13)

**Reply by johnmarksilly** [reply]
Ah I think I get what happened. It is stupid. The only thing that made me think it lasted longer than it should have was that the thumbnail was showing the car from this morning. I assume it was just the latest snapshot. Good god sorry for the strife.

Thanks Nick!
